pkg/provider/ngrok/client.go

```diff
@@ -0,0 +1,329 @@
+func (c *client) DeleteSecret(ctx context.Context, ref esv1.PushSecretRemoteRef) error {
+	err := c.verifyVaultNameStillMatchesID(ctx)
+	if errors.Is(err, errVaultDoesNotExist) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	secret, err := c.getSecretByVaultIDAndName(ctx, c.getVaultID(), ref.GetRemoteKey())
+	if errors.Is(err, errVaultDoesNotExist) || errors.Is(err, errVaultSecretDoesNotExist) {
+		// If the secret or vault do not exist, we can consider it deleted.
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
+
+	if secret == nil {
+		return nil
+	}
+
+	return c.secretsClient.Delete(ctx, secret.ID)
+}
+
+func (c *client) Validate() (esv1.ValidationResult, error) {
+	// Validate the client can list secrets with a timeout. If we
+	// can list secrets, we assume the client is valid(API keys, URL, etc.)
+	iter := c.secretsClient.List(nil)
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	for iter.Next(ctx) {
+		return esv1.ValidationResultReady, nil
+	}
+
+	if iter.Err() != nil {
+		return esv1.ValidationResultError, fmt.Errorf("store is not allowed to list secrets: %w", iter.Err())
+	}
+
+	return esv1.ValidationResultReady, nil
+}
+
+func (c *client) GetSecret(ctx context.Context, _ esv1.ExternalSecretDataRemoteRef) ([]byte, error) {
+	// Implementation for getting a secret from ngrok
+	return nil, errWriteOnlyOperations
+}
+
+func (c *client) GetSecretMap(ctx context.Context, _ esv1.ExternalSecretDataRemoteRef) (map[string][]byte, error) {
+	// Implementation for getting a map of secrets from ngrok
+	return nil, errWriteOnlyOperations
+}
+
+func (c *client) GetAllSecrets(ctx context.Context, _ esv1.ExternalSecretFind) (map[string][]byte, error) {
+	// Implementation for getting all secrets from ngrok
+	return nil, errWriteOnlyOperations
+}
+
+func (c *client) Close(_ context.Context) error {
+	return nil
+}
+
+func (c *client) verifyVaultNameStillMatchesID(ctx context.Context) error {
+	if c.getVaultID() == "" {
+		return c.refreshVaultID(ctx)
+	}
+
+	vault, err := c.vaultClient.Get(ctx, c.getVaultID())
+	if err != nil || vault == nil || vault.Name != c.vaultName {
+		return c.refreshVaultID(ctx)
+	}
+
+	c.setVaultID(vault.ID)
+	return nil
+}
+
+// getVaultID safely retrieves the current vault ID.
+func (c *client) getVaultID() string {
+	c.vaultIDMu.RLock()
+	defer c.vaultIDMu.RUnlock()
+	return c.vaultID
+}
+
+// setVaultID safely sets the vault ID.
+func (c *client) setVaultID(vaultID string) {
+	c.vaultIDMu.Lock()
+	defer c.vaultIDMu.Unlock()
+	c.vaultID = vaultID
+}
+
+func (c *client) refreshVaultID(ctx context.Context) error {
+	v, err := c.getVaultByName(ctx, c.vaultName)
+	if err != nil {
+		return fmt.Errorf("failed to refresh vault ID: %w", err)
+	}
+
+	c.setVaultID(v.ID)
+	return nil
+}
+
+func (c *client) getVaultByName(ctx context.Context, name string) (*ngrok.Vault, error) {
+	iter := c.vaultClient.List(nil)
```

Comment:
you use List(nil) everywhere which indicates no paging AFAIK. However to me it seems like that is not quite right. For things such as Validation or in general you want a sane default page size right? is that defaulted from client side? whats the behavior here?

Reply:
There is paging that will happen by default if not supplied.[ If limit is not supplied, it defaults to 100 items](https://ngrok.com/docs/api/#pagination).  [Here](https://github.com/ngrok/ngrok-api-go/blob/v7.6.0/vaults/client.go#L230-L232) is where that defaulting happens in our API client if you pass nil.